tso: pd nit pointer panic in tso dispatcher

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?

<!-- If possible, provide a recipe for reproducing the error. -->

### What did you expect to see?
normal 
### What did you see instead?
[panic](https://github.com/tikv/pd/actions/runs/4373197579/jobs/7651070154)
```
2023-03-09T10:05:14.8268865Z [2023/03/09 10:05:13.501 +00:00] [INFO] [client.go:290] ["[pd] create pd client with endpoints"] [pd-address="[https://127.0.0.1:36315]"]
2023-03-09T10:05:14.8269328Z [2023/03/09 10:05:13.554 +00:00] [INFO] [base_client.go:480] ["[pd] switch leader"] [new-leader=https://127.0.0.1:36315] [old-leader=]
2023-03-09T10:05:14.8269675Z [2023/03/09 10:05:13.554 +00:00] [INFO] [base_client.go:156] ["[pd] init cluster id"] [cluster-id=7208485466879229826]
2023-03-09T10:05:14.8270049Z [2023/03/09 10:05:13.555 +00:00] [INFO] [tso_request_dispatcher.go:287] ["[pd/tso] tso dispatcher created"] [dc-location=global]
2023-03-09T10:05:14.8270446Z [2023/03/09 10:05:13.555 +00:00] [INFO] [client.go:290] ["[pd] create pd client with endpoints"] [pd-address="[https://127.0.0.1:36315]"]
2023-03-09T10:05:14.8270786Z [2023/03/09 10:05:13.562 +00:00] [ERROR] [tso_request_dispatcher.go:507] ["[pd/tso] update connection contexts failed"] [dc=global] [error="rpc error: code = Canceled desc = context canceled"]
2023-03-09T10:05:14.8271151Z [2023/03/09 10:05:13.562 +00:00] [INFO] [tso_request_dispatcher.go:305] ["[pd/tso] exit tso dispatcher"] [dc-location=global]
2023-03-09T10:05:14.8271355Z panic: runtime error: invalid memory address or nil pointer dereference
2023-03-09T10:05:14.8271548Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1eedf59]
2023-03-09T10:05:14.8271566Z 
2023-03-09T10:05:14.8271675Z goroutine 348500 [running]:
2023-03-09T10:05:14.8271933Z github.com/tikv/pd/client.(*client).tryResourceManagerConnect(0xc007054720?, {0x554f930, 0xc00a167bd0}, 0xc0057c1e58)
2023-03-09T10:05:14.8272146Z 	/home/runner/work/pd/pd/client/resource_manager_client.go:354 +0x119
2023-03-09T10:05:14.8272414Z github.com/tikv/pd/client.(*client).handleResourceTokenDispatcher(0xc015a9ae10, {0x554f930, 0xc00a167bd0}, 0xc015bdd300)
2023-03-09T10:05:14.8272628Z 	/home/runner/work/pd/pd/client/resource_manager_client.go:278 +0xdb
2023-03-09T10:05:14.8272820Z created by github.com/tikv/pd/client.(*client).createTokenDispatcher
2023-03-09T10:05:14.8273015Z 	/home/runner/work/pd/pd/client/resource_manager_client.go:265 +0x2cf
2023-03-09T10:05:14.8273162Z FAIL	github.com/tikv/pd/tests/client	230.444s
2023-03-09T10:05:14.8273243Z FAIL
```
### What version of PD are you using (`pd-server -V`)?
master

--- a/client/grpcutil/grpcutil.go
+++ b/client/grpcutil/grpcutil.go
@@ -74,7 +74,7 @@ func BuildForwardContext(ctx context.Context, addr string) context.Context {
 }
 
 // GetOrCreateGRPCConn returns the corresponding grpc client connection of the given addr.
-// Returns the old one if's already  existed in the clientConns; otherwise creates a new one and returns it.
+// Returns the old one if's already existed in the clientConns; otherwise creates a new one and returns it.
 func GetOrCreateGRPCConn(ctx context.Context, clientConns *sync.Map, addr string, tlsCfg *tlsutil.TLSConfig, opt ...grpc.DialOption) (*grpc.ClientConn, error) {
 	conn, ok := clientConns.Load(addr)
 	if ok {

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -366,12 +366,11 @@ func (c *client) tryResourceManagerConnect(ctx context.Context, connection *reso
 		stream rmpb.ResourceManager_AcquireTokenBucketsClient
 	)
 	for i := 0; i < maxRetryTimes; i++ {
-		cctx, cancel := context.WithCancel(ctx)
 		cc, err := c.resourceManagerClient()
 		if err != nil {
-			cancel()
-			return err
+			continue
 		}
+		cctx, cancel := context.WithCancel(ctx)
 		stream, err = cc.AcquireTokenBuckets(cctx)
 		if err == nil && stream != nil {
 			connection.cancel = cancel

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -67,11 +67,11 @@ func (c *client) gRPCErrorHandler(err error) {
 
 // ListResourceGroups loads and returns all metadata of resource groups.
 func (c *client) ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error) {
-	req := &rmpb.ListResourceGroupsRequest{}
 	cc, err := c.resourceManagerClient()
 	if err != nil {
 		return nil, err
 	}
+	req := &rmpb.ListResourceGroupsRequest{}
 	resp, err := cc.ListResourceGroups(ctx, req)
 	if err != nil {
 		c.gRPCErrorHandler(err)
@@ -85,12 +85,12 @@ func (c *client) ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup,
 }
 
 func (c *client) GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error) {
-	req := &rmpb.GetResourceGroupRequest{
-		ResourceGroupName: resourceGroupName,
-	}
 	cc, err := c.resourceManagerClient()
 	if err != nil {
 		return nil, err
+	}
+	req := &rmpb.GetResourceGroupRequest{
+		ResourceGroupName: resourceGroupName,
 	}
 	resp, err := cc.GetResourceGroup(ctx, req)
 	if err != nil {
@@ -113,14 +113,14 @@ func (c *client) ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.Resour
 }
 
 func (c *client) putResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup, typ actionType) (string, error) {
-	req := &rmpb.PutResourceGroupRequest{
-		Group: metaGroup,
-	}
-	var resp *rmpb.PutResourceGroupResponse
 	cc, err := c.resourceManagerClient()
 	if err != nil {
 		return "", err
 	}
+	req := &rmpb.PutResourceGroupRequest{
+		Group: metaGroup,
+	}
+	var resp *rmpb.PutResourceGroupResponse
 	switch typ {
 	case add:
 		resp, err = cc.AddResourceGroup(ctx, req)
@@ -139,12 +139,12 @@ func (c *client) putResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceG
 }
 
 func (c *client) DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error) {
-	req := &rmpb.DeleteResourceGroupRequest{
-		ResourceGroupName: resourceGroupName,
-	}
 	cc, err := c.resourceManagerClient()
 	if err != nil {
 		return "", err
+	}
+	req := &rmpb.DeleteResourceGroupRequest{
+		ResourceGroupName: resourceGroupName,
 	}
 	resp, err := cc.DeleteResourceGroup(ctx, req)
 	if err != nil {

--- a/client/resource_manager_client.go
+++ b/client/resource_manager_client.go
@@ -50,11 +50,12 @@ type ResourceManagerClient interface {
 }
 
 // resourceManagerClient gets the ResourceManager client of current PD leader.
-func (c *client) resourceManagerClient() rmpb.ResourceManagerClient {
-	if cc, err := c.svcDiscovery.GetOrCreateGRPCConn(c.GetLeaderAddr()); err == nil {
-		return rmpb.NewResourceManagerClient(cc)
+func (c *client) resourceManagerClient() (rmpb.ResourceManagerClient, error) {
+	cc, err := c.svcDiscovery.GetOrCreateGRPCConn(c.GetLeaderAddr())
+	if err != nil {
+		return nil, err
 	}
-	return nil
+	return rmpb.NewResourceManagerClient(cc), nil
 }
 
 // gRPCErrorHandler is used to handle the gRPC error returned by the resource manager service.
@@ -67,7 +68,11 @@ func (c *client) gRPCErrorHandler(err error) {
 // ListResourceGroups loads and returns all metadata of resource groups.
 func (c *client) ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error) {
 	req := &rmpb.ListResourceGroupsRequest{}
-	resp, err := c.resourceManagerClient().ListResourceGroups(ctx, req)
+	cc, err := c.resourceManagerClient()
+	if err != nil {
+		return nil, err
+	}
+	resp, err := cc.ListResourceGroups(ctx, req)
 	if err != nil {
 		c.gRPCErrorHandler(err)
 		return nil, errs.ErrClientListResourceGroup.FastGenByArgs(err.Error())
@@ -83,7 +88,11 @@ func (c *client) GetResourceGroup(ctx context.Context, resourceGroupName string)
 	req := &rmpb.GetResourceGroupRequest{
 		ResourceGroupName: resourceGroupName,
 	}
-	resp, err := c.resourceManagerClient().GetResourceGroup(ctx, req)
+	cc, err := c.resourceManagerClient()
+	if err != nil {
+		return nil, err
+	}
+	resp, err := cc.GetResourceGroup(ctx, req)
 	if err != nil {
 		c.gRPCErrorHandler(err)
 		return nil, &errs.ErrClientGetResourceGroup{ResourceGroupName: resourceGroupName, Cause: err.Error()}
@@ -103,34 +112,41 @@ func (c *client) ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.Resour
 	return c.putResourceGroup(ctx, metaGroup, modify)
 }
 
-func (c *client) putResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup, typ actionType) (str string, err error) {
+func (c *client) putResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup, typ actionType) (string, error) {
 	req := &rmpb.PutResourceGroupRequest{
 		Group: metaGroup,
 	}
 	var resp *rmpb.PutResourceGroupResponse
+	cc, err := c.resourceManagerClient()
+	if err != nil {
+		return "", err
+	}
 	switch typ {
 	case add:
-		resp, err = c.resourceManagerClient().AddResourceGroup(ctx, req)
+		resp, err = cc.AddResourceGroup(ctx, req)
 	case modify:
-		resp, err = c.resourceManagerClient().ModifyResourceGroup(ctx, req)
+		resp, err = cc.ModifyResourceGroup(ctx, req)
 	}
 	if err != nil {
 		c.gRPCErrorHandler(err)
-		return str, err
+		return "", err
 	}
 	resErr := resp.GetError()
 	if resErr != nil {
-		return str, errors.Errorf("[resource_manager] %s", resErr.Message)
+		return "", errors.Errorf("[resource_manager] %s", resErr.Message)
 	}
-	str = resp.GetBody()
-	return
+	return resp.GetBody(), nil
 }
 
 func (c *client) DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error) {
 	req := &rmpb.DeleteResourceGroupRequest{
 		ResourceGroupName: resourceGroupName,
 	}
-	resp, err := c.resourceManagerClient().DeleteResourceGroup(ctx, req)
+	cc, err := c.resourceManagerClient()
+	if err != nil {
+		return "", err
+	}
+	resp, err := cc.DeleteResourceGroup(ctx, req)
 	if err != nil {
 		c.gRPCErrorHandler(err)
 		return "", err
@@ -351,7 +367,12 @@ func (c *client) tryResourceManagerConnect(ctx context.Context, connection *reso
 	)
 	for i := 0; i < maxRetryTimes; i++ {
 		cctx, cancel := context.WithCancel(ctx)
-		stream, err = c.resourceManagerClient().AcquireTokenBuckets(cctx)
+		cc, err := c.resourceManagerClient()
+		if err != nil {
+			cancel()
+			return err
+		}
+		stream, err = cc.AcquireTokenBuckets(cctx)
 		if err == nil && stream != nil {
 			connection.cancel = cancel
 			connection.ctx = cctx

